### PR TITLE
fix(amazon): Update load balancer validations to match user expectations

### PR DESCRIPTION
**Pull request by caseyhebebrand**
Addressing validation issues on the create/edit load balancer user flow: 

1. Form validation for health check values was invoked on the user string, causing the validation to fail. ("Must be a number")
2. Inconsistent with [AWS](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-elasticloadbalancingv2-targetgroup.html) rules. The summary is below, but more details can be found in the doc I linked. 

**HealthCheckTimeoutSeconds**
For target groups with TCP/TLS protocols - it must be 6 seconds for HTTP healthchecks and 10 seconds for TCP/HTTPS healthchecks.

For target groups with HTTP/HTTPS protocols, the range of values 2-120 should be acceptable.

**HealthCheckIntervalSeconds**
For TCP healthchecks, the interval must be 10 or 30 seconds.

### CODE REVIEW

**Comment by christopherthielen** [comment]
It might make sense to migrate these fields to `<SpelNumberInput>` instead.

This would also be a great opportunity to migrate this old `validate()` method to `FormValidation` and prove out the recent-ish `spelAware` validation methods from `validation.ts`

**Comment by caseyhebebrand** [comment]
> For the cases where there are only a couple of specific values accepted, why not a radio input? For the cases where only one value is acceptable, why make people choose at all vs. just showing the value that will be used?

That is a great point. I will just show the chosen value for the two cases where only one value is acceptable. For the case where there are 2 options I will think a little more about the best user experience for this - either a radio or a tooltip and the same input as the other cases.

**Comment by caseyhebebrand** [comment]
@erikmunson So I took a similar approach to what you were suggesting. For the strict use cases with only 1 valid option, I render a disabled input with the valid value and a `HelpField` explaining why they cannot select another value. For those with only a couple options I show a `HelpField` to explain which options are valid.

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import { get } from 'lodash';
+import { Validators, robotToHuman, spelNumberCheck } from '@spinnaker/core';
+import { ITargetGroup } from 'amazon/domain';
+
+export const isNameInUse = (
+  existingNames: { [account: string]: { [region: string]: string[] } },
+  credentials: string,
+  region: string,
+) => (name: string) =>
+  get(existingNames, [credentials, region], []).includes(name.toLowerCase())
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts]: Oh wow, TIL that the `path` argument to `get` can be an array!

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+export const isNameLong = (appNameLength: number) => (name: string) =>
+  name.length < 32 - appNameLength
+    ? null
+    : 'Target group name is automatically prefixed with the application name and cannot exceed 32 characters in length.';
+
+export const isDuplicateName = (duplicateGroups: string[]) => (name: string) =>
+  !duplicateGroups.includes(name) ? null : 'Duplicate target group name in this load balancer.';
+
+export const isValidTimeout = (targetGroup: ITargetGroup) => (value: string) => {
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts]: should this be `(value: number)` now that you've migrated to NumberInput?

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+};
+
+export const isValidHealthCheckInterval = (targetGroup: ITargetGroup) => (value: string) =>
+  targetGroup.healthCheckProtocol !== 'TCP' ||
+  (targetGroup.healthCheckProtocol === 'TCP' &&
+    (Number.parseInt(value, 10) === 10 || Number.parseInt(value, 10) === 30))
+    ? null
+    : 'TCP health checks only support 10s and 30s intervals';
+
+export const checkBetween = (fieldName: string, min: number, max: number) => (value: string) => {
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts]: This one could probably be moved to `@spinnaker/core` validators as it's a common (non-amazon specific) case.  We could use it inside `NumberInput` as well.

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import { get } from 'lodash';
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts]: Awesome work with the validators.  The factory functions can be typed to return `IValidator`

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+          'us-east-1': ['targetgroup2'],
+        },
+      };
+      const actual = Validators.isNameInUse(existingGroups, 'test', 'us-east-1')(mockTargetGroup.name);
+      expect(actual).toEqual(null);
+    });
+  });
+
+  describe('of name length', () => {
+    it('returns error if the name is >32 additional characters', () => {
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts]: might be worth explaining somehow in the `it` block that the application name is prepended

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+      const actual = Validators.isNameLong(12)(`${mockTargetGroup.name}areallylongname`);
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts]: minor testing nit: In assertions like this, it's easier to understand the intention of the test if you use literal values instead of references, even if it means more "duplication".  For example, I think this assertion would be easier to grok if it was:

```suggestion
      const actual = Validators.isNameLong('application13'.length)('areallylongtargetgroupname');
```

This way, you don't have to wonder what `12` means, nor what is in `mockTargetGroup.name`

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  get(existingNames, [credentials, region], []).includes(name.toLowerCase())
+    ? `There is already a target group in ${credentials}:${region} with that name.`
+    : null;
+
+export const isNameLong = (appNameLength: number) => (name: string) =>
+  name.length < 32 - appNameLength
+    ? null
+    : 'Target group name is automatically prefixed with the application name and cannot exceed 32 characters in length.';
+
+export const isDuplicateName = (duplicateGroups: string[]) => (name: string) =>
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts]: This validator already exists in core as `valueUnique`

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    const error =
+      Validators.minValue(min)(sanitizedField, robotToHuman(fieldName)) ||
+      Validators.maxValue(max)(sanitizedField, robotToHuman(fieldName));
+
+    return error;
+  }
+  return null;
+};
+
+export const spelNumber = (value: number | string) => spelNumberCheck(value);
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts]: `spelNumberCheck` was created before the validation api was finalized.  I'd prefer to create a (non-spel aware) `isNumber` validator in core and use that, deprecating `spelNumberCheck` entirely.  `isNumber` could then be composed into the other number validators (`minValue`, `maxValue`, `between`)

File: app/scripts/modules/amazon/src/loadBalancer/configure/application/TargetGroups.tsx [hunk 9]
```diff
@@ -357,8 +341,7 @@ export class TargetGroups extends React.Component<ITargetGroupsProps, ITargetGro
                           <span className="wizard-pod-content">
                             <label>Path </label>
                             <SpInput
-                              className="form-control input-sm inline-text"
-                              error={tgErrors.healthCheckPath}
+                              // error={tgErrors.healthCheckPath}
```
Comment (christopherthielen) [inline app/scripts/modules/amazon/src/loadBalancer/configure/application/TargetGroups.tsx]: remove?

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+describe('Target Group validators', () => {
+  describe('of used names', () => {
+    it('returns an error if the name exists already', () => {
+      const existingGroups = {
+        test: {
+          'us-east-1': ['targetgroup'],
+        },
+      };
+      const actual = Validators.isNameInUse(existingGroups, 'test', 'us-east-1')(mockTargetGroup.name);
+      const expected = 'There is already a target group in test:us-east-1 with that name.';
```
Comment (erikmunson) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts]: is there value in asserting the exact error message in these tests (given that they'll likely just get copy-pasted across files when changed), as opposed to only asserting that the validator returns _some_ error message when it's expected to? e.g. `expect(isString(actual))`?

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+  region: 'us-east-1',
+  type: '',
+};
+
+describe('Target Group validators', () => {
+  describe('of used names', () => {
+    it('returns an error if the name exists already', () => {
+      const existingGroups = {
+        test: {
+          'us-east-1': ['targetgroup'],
```
Comment (erikmunson) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts]:
```suggestion
          'us-east-1': [mockTargetGroup.name],
```

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+
+    it('returns null if the name is < 32', () => {
+      const actual = Validators.isNameLong(12)(mockTargetGroup.name);
+      expect(actual).toEqual(null);
+    });
+  });
+
+  describe('of duplicate names across the load balancer', () => {
+    it('returns an error if the name exists', () => {
+      const actual = Validators.isDuplicateName(['targetgroup'])(mockTargetGroup.name);
```
Comment (erikmunson) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts]:
```suggestion
      const actual = Validators.isDuplicateName([mockTargetGroup.name])(mockTargetGroup.name);
```

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+      expect(actual).toEqual(expected);
+    });
+
+    it('should be 6s and is valid', () => {
+      const actual = Validators.isValidTimeout(tg)('6');
+      expect(actual).toEqual(null);
+    });
+
+    it('should have a 10s timeout', () => {
+      const actual = Validators.isValidTimeout({ ...tg, healthCheckProtocol: 'HTTPS' })('9');
```
Comment (erikmunson) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts]: i found this double-override (first in the definition of `tg` and then inline in the `it` blocks) a bit hard to follow. Thoughts on just defining multiple variables at the top with friendlier names? e.g.
```ts
    const httpTargetGroup = {
      ...mockTargetGroup,
      protocol: 'TCP',
      healthCheckProtocol: 'HTTP',
    };

    const httpsTargetGroup = {
      ...mockTargetGroup,
      protocol: 'TCP',
      healthCheckProtocol: 'HTTPS',
    };

```

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+      const actual = Validators.isValidTimeout(mockTargetGroup)('10');
+      expect(actual).toEqual(null);
+    });
+  });
+
+  describe('of health check interval', () => {
+    const tg = {
+      ...mockTargetGroup,
+      healthCheckProtocol: 'TCP',
+    };
```
Comment (erikmunson) [inline app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts]: similar to above, `tcpTargetGroup` might be a bit friendlier.

File: app/scripts/modules/amazon/src/loadBalancer/configure/application/TargetGroups.tsx [hunk 15]
```diff
@@ -369,27 +352,42 @@ export class TargetGroups extends React.Component<ITargetGroupsProps, ITargetGro
...
+                                }
+                              />
+                            ) : (
+                              <SpelNumberInput
+                                error={tgErrors.healthCheckTimeout}
+                                required={true}
+                                value={targetGroup.healthCheckTimeout}
+                                min={2}
+                                max={120}
+                                onChange={(value: string) =>
```
Comment (erikmunson) [inline app/scripts/modules/amazon/src/loadBalancer/configure/application/TargetGroups.tsx]: I noticed that everywhere you're using `SpelNumberInput` (for both ALB/NLB) the type of this callback has been narrowed to `string` — what's the context behind that? Is there something downstream of these callbacks that gets angry if you let the `number | string` type flow through?

I only ask because in theory (not always in practice) it's preferable to let the types that are already defined flow, and in this case it seems like the narrowed `string` type also isn't entirely accurate re: what will actually happen at runtime?

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts (hunk 10 above)
Comment (caseyhebebrand) [reply app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts]: Its more of a nit-picky value to me. It brings value if we care about the specific language being used. I come from a world of very strict language requirements, therefore I have a habit of this, but I am happy to change this if you feel it is too strict.

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts (hunk 13 above)
Comment (caseyhebebrand) [reply app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.spec.ts]: No strong opinions. I always lean toward repeating code as little as possible, but Chris expressed a similar concern as you so I will go make the references a bit more readable.

File: app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts (hunk 7 above)
Comment (caseyhebebrand) [reply app/scripts/modules/amazon/src/loadBalancer/configure/common/targetGroupValidators.ts]: Great, will convert to that!
